File: 《算法4th》/实现/1.4 背包.java
/*
API - Bag
void add(T temp)
boolean isEmpty()
int size()
*/

import java.util.Iterator;

public class Bag<T> implements Iterable<T>{
	private class Node{
		T value;
		Node next;
	}
	private Node first;
	
	public void add(T temp){
		Node oldfirst = first;
		first = new Node();
		first.value = temp;
		first.next = oldfirst;
	}
	
	//	hasNext() next() remove()供用例的foreach语法使用
	public Iterator<T> iterator(){ return new ListIterator(); }
	Private class ListIterator implements Iterator<T>{
		private Node current = first;
		public boolean hasNext(){ return current != NULL }
		public T next(){
			T rst = current.value;
			current = current.next;
			return rst;
		}
		public void remove(){}
	}
}

File: 《算法4th》/实现/1.4 背包.cpp
/*
API - Bag
void push(T temp)
*/

template <class T> 
class Stack {
public:
	void push(T temp)(){
		Node oldfirst = first;
		first = new Node();
		first.value = temp;
		first.next = oldfirst;
	}
private:
	struct Node{
		T value;
		Node next;
	}
private:
	Node first;
	
	
	
	//	iterator()之后实现
}